### PR TITLE
Release 3.1.0 master PR

### PR DESCRIPTION
**Pull request by mfrancis95**
Starting this to serve as a master PR for release 3.1.0. Locking in 7f25bf6fbd0b6917912db86d5e321886375247aa as the commit 3.1 is based on. I figured we can use this PR for a bit to collect some more changelog additions before going live.

There are simply way too many changes to try to sift through though, so eventually we should reach some level of comfort with what we have and consider augmenting the changelog post-release if we find additional key changes that we want captured.

### CODE REVIEW

**Comment by turol** [comment]
Things to mention:
* improved portability by allowing disable SDL extra libs
* various windows midi improvements @ceski-1 @rfomin can you elaborate all the things that were fixed?
* -display command line parameter
* Fix mouse event causing unwanted key press d66009a89528c4a8686319a7e6b1975383187080
* pcsound fixes 080ed4ee4324fa873b4049f5cda8749d9bd801c7
* cleaned up abuses of extern
* cleaned up function pointer correctness in heretic, hexen, might improve portability
* CI improvements: mingw build, cppcheck, linter check
* support for extra mouse buttons (maybe, not sure if existed before and was just fixed) 6e363ea1bad43a38b85882e89a9386566b512bf4
* was gameversion 1.2 support new or just fixed?
* new icons
* developer only debug things (mempool disable, linked list infinite loop checking)
* windows non-latin fixes #1481 #1486 #1463
* general compatibility fixes with new compilers (includes, warnings)
* use artifact mouse button (heretic, hexen) #1459 
* inventory mouse buttons heretic #1256 hexen #1458 
* french wads #1418
* mouse button run #1415 
* hexen coop demo playback #1382 
* support autoloading gog.com releases #1338
* python 2 deprecated, python 3 required #1330 
*

**Comment by rfomin** [comment]
* Redesign of native MIDI support in Windows (authors @ceski-1 and @rfomin):
  - Removing the `midiproc.exe` program.
  - Correct pause/resume.
  - Added support for SysEx messages.
  - Correct reset of MIDI devices using SysEx messages (by @ceski-1).
  - Implement a "capital tone fallback" emulation (by @ceski-1).
  - Add full EMIDI support, looping points (Final Fantasy and RPG Maker) (by @ceski-1).
  - Add MIDI compatibility levels. Emulate Vanilla Doom's DMX MPU-401 mode, Microsoft GS Wavetable Synth or send all messages to MIDI device. Documentation https://gist.github.com/ceski-1/86b7405a66ff67c42a788000edc3b625 (by @ceski-1).
* FluidSynth support implemented (by @mikeday0).
* Support for non-Latin file names in Windows (by @rfomin).
* Support for modern gamepads (by @mikeday0).

Hopefully @ceski-1 and @mikeday0 will add more details.

A few issues:
1. The current `chocpkg` system does not support FluidSynth. We can build a release for Windows using MSYS2.
2. I think we should add support for at least analog turning for the gamepad (analog movement is more controversial).

**Comment by mikeday0** [comment]
* FluidSynth and gamepad contributions are already mentioned in NEWS.md, so I think we're covered on that.
* Properly handle orphan carriage returns when parsing deh files (by @mikeday0).
* Allow simultaneous PC speaker emulation and OPL emulation (by @mikeday0).

Everything else I did was already mentioned in turol's comment above.

> The current `chocpkg` system does not support FluidSynth. We can build a release for Windows using MSYS2 in GitHub CI.

If we go the GitHub CI MSYS2 route, I suggest waiting for the new release of SDL_Mixer. (I saw somewhere they were doing one this month) It will have a fix for the annoying FLAC looping crash.

> I think we should add support for at least analog turning for the gamepad (analog movement is more controversial).

I've got this mostly done, just need to polish. I will submit a PR this weekend.

**Comment by ceski-1** [comment]
Some additional details for Windows MIDI support. Feel free to combine the similar items with what @rfomin already listed:
- Emulates DMX MPU-401 feature set by default. Adjust with `winmm_complevel` for modern MIDIs.
- Device is reset on startup, shutdown, and between songs using a "GM System On" SysEx message by default. Adjust with `winmm_reset_type`.
- Added `winmm_reset_delay` for legacy hardware devices (e.g. Roland SC-55). Fixes https://github.com/chocolate-doom/chocolate-doom/issues/1547.
- Added "capital tone fallback" emulation in GS mode, similar to early versions of the Roland SC-55.
- Music volume curve matches SDL's implementation. This is closer to vanilla but may seem louder than previous versions of Chocolate Doom.
- A [preferred device can be selected](https://github.com/chocolate-doom/chocolate-doom/assets/56656010/c0eace33-03ea-4214-8234-11ca20a4471a) from a list of all devices in setup.

**Comment by mfrancis95** [comment]
I'm starting to slowly add what's been mentioned. Feel free to use the suggestion feature to add your lines or add commits to my branch.

**Comment by fabiangreffrath** [comment]
Please mention the two security fixes that were factor out into 3.0.1, but are if course also fixed in 3.1.0.

**Comment by fabiangreffrath** [comment]
> Please mention the two security fixes that were factor out into 3.0.1, but are if course also fixed in 3.1.0.

One of them was assigned a CVE id, that's important to mention in the changelog.

**Comment by mfrancis95** [comment]
@fabiangreffrath 3.0.1 added.

**Comment by mfrancis95** [comment]
Added @fragglet 's NEWS.md additions from last year. @turol judging by git blame it looks like the enum for v1.2 was added pre-3.0.0 but the version didn't gain any actual semblance of emulation until post-3.0.0.

**Comment by mfrancis95** [comment]
Added more from @turol 's list. Can you expand on `improved portability by allowing disable SDL extra libs` and provide a commit hash if you can? It's to help credit people and understand the context. Otherwise, it would be preferable to use the suggestion feature to add those in.

Does anyone know who were all the folks who contributed to the new icons? I see several people possibly involved and it looks like at least one of them has a deleted GitHub account. Just want to make sure credit is given where it's due.

**Comment by fabiangreffrath** [comment]
> Does anyone know who were all the folks who contributed to the new icons? I see several people possibly involved and it looks like at least one of them has a deleted GitHub account. Just want to make sure credit is given where it's due.

The person who left Github was the one who made the current Crispy icons. Their name is Philip Kiwan, but that's all I know unfortunately.

**Comment by fabiangreffrath** [comment]
Oh, wow, apparently the original contributor of the chocolate bar style icons that are currently used for Chocolate has also disappeared. According to this commit, their user name was @clovehoof:
https://github.com/chocolate-doom/chocolate-doom/commit/709862f79010c66686c88574dd8f2a3f1b267180

**Comment by turol** [comment]
@mfrancis95 explanations in the PRs:
https://github.com/chocolate-doom/chocolate-doom/pull/1489
https://github.com/chocolate-doom/chocolate-doom/pull/1480

Also re: mingw CI, I only copied that from Crispy Doom so most of the credit probably belongs to @fabiangreffrath

**Comment by fabiangreffrath** [comment]
> Also re: mingw CI, I only copied that from Crispy Doom so most of the credit probably belongs to @fabiangreffrath

Nope, not my work. Please credit @rfomin for this.

File: NEWS.md
Comment (fabiangreffrath) [inline NEWS.md:105]: It is now even possible to bind mouse buttons to turn left/right in all games:
https://github.com/chocolate-doom/chocolate-doom/commit/331aed42b4490388d0d60867a731b7b7e63e8f3e

**Comment by rfomin** [comment]
> > Also re: mingw CI, I only copied that from Crispy Doom so most of the credit probably belongs to @fabiangreffrath
> 
> Nope, not my work. Please credit @rfomin for this.

Speaking of mingw CI, I can modify it to publish a release on GitHub like Crispy does.

Pros:
* Newer versions of the compiler and tools
*  64-bit build
*  FluidSynth support

Cons:
* We will lose support for Windows XP

**Comment by fabiangreffrath** [comment]
> * We will lose support for Windows XP

I am all for it! Even Microsoft itself gave up support for Windows XP some 10 years ago.

**Comment by mfrancis95** [comment]
> @mfrancis95 explanations in the PRs: #1489 #1480

Thanks @turol. Added.

> Speaking of mingw CI, I can modify it to publish a release on GitHub like Crispy does.

@rfomin how would you safeguard against having it publish a release for every single commit? I would think we'd only want to have it do this every now and then (but definitely way more frequently than 6 years!)

> @mfrancis95 There are several open PRs ready for review. Do you want to keep master frozen or can I merge them?

I personally really want to just focus on getting this out the door as-is because more PRs = more changelog additions, but I know that @rfomin and @mikeday0 might be looking to get some changes in, so let me know how you feel about saving some work for `3.2.0`? My open PRs can be ignored -- I still want to test some of mine and I have no issues saving them for the next release.

**Comment by ceski-1** [comment]
> let me know how you all feel about saving some work for 3.2.0?

Will there be a return to regular Chocolate Doom releases? My PR can wait, but better gamepad support (that includes analog turning/movement) and the SDL_mixer FLAC bug are mentioned often by users.

**Comment by rfomin** [comment]
> @rfomin how would you safeguard against having it publish a release for every single commit?

A release will be published when a new tag is pushed to the repository, this is how it works for the last few Crispy Doom releases. I'll add this code to `main.yml`: https://github.com/fabiangreffrath/crispy-doom/blob/99416f4be87f1a5eb6e169c670280627284735a9/.github/workflows/msys2.yml#L79-L86

> let me know how you all feel about saving some work for 3.2.0?

I agree with @ceski-1, updating the gamepad https://github.com/chocolate-doom/chocolate-doom/pull/1629 is important. If we decide to use GitHub CI for releases, we should also wait for the SDL_mixer update in MSYS2 to fix the FLAC bug.

**Comment by fabiangreffrath** [comment]
> I have no issues saving them for the next release.

I don't think we are in a hurry. We have waited for so long, that it doesn't really make a difference to wait a bit more if this means we can take the opportunities for some real improvements. For me, I'd like to avoid releasing 3.1 today, after 6 year of silence, and 3.2 one week later.

**Comment by turol** [comment]
Need to mention
#1624
#1630 
#1631

**Comment by ceski-1** [comment]
Added Strife PC speaker sound support: https://github.com/chocolate-doom/chocolate-doom/pull/1632

This PR is fairly important and should be included: https://github.com/chocolate-doom/chocolate-doom/pull/1637
@turol Please take a look if you have time. Your reviews are always appreciated.

**Comment by turol** [comment]
The lines are inconsistent. Some are one long line, others have been split up. Which one should we use?

**Comment by ceski-1** [comment]
> The lines are inconsistent. Some are one long line, others have been split up. Which one should we use?

Based on [this](https://github.com/chocolate-doom/chocolate-doom/blob/master/NEWS.md?plain=1), it appears to follow an 80-column wide format, so I'd go with that for consistency.

**Comment by turol** [comment]
Are we still missing something?

Should be (manually) squashed to remove \@:s (so people don't get pinged when someone pushes)

There was something about Windows builds not having up to date libraries. Has this been fixed? If not, what needs to be fixed?

**Comment by fabiangreffrath** [comment]
We'll need to update SDL2_Mixer to 2.28 in chocpkg.

**Comment by turol** [comment]
Is something blocking that?

**Comment by ceski-1** [comment]
> Are we still missing something?

Some changelog items seem to be missing still.

**Comment by turol** [comment]
> Some changelog items seem to be missing still.

Which ones? Some of the PRs I asked about were commented as already covered.

**Comment by ceski-1** [comment]
> Which ones? Some of the PRs I asked about were commented as already covered.

<details><summary>Changelog additions</summary>

```md

## 3.1.0 (2023-??-??)

### General
* Allow simultaneous PC speaker emulation and OPL emulation (thanks Michael Day).
* Redesign of native MIDI support in Windows (thanks ceski and Roman Fomin):
    - Emulates DMX MPU-401 feature set by default, change with `winmm_complevel`.
    - MIDI device reset between songs using "GM System On", change with `winmm_reset_type`.
    - Configurable delay after reset for legacy MIDI hardware (e.g. Roland SC-55), change with `winmm_reset_delay`.

### Bug fixes
* Properly handle orphan carriage returns when parsing deh files (thanks Michael Day).

### Doom
* Mouse buttons are now bindable to turn left/right (thanks Fabian).

### Heretic
* Mouse buttons are now bindable to turn left/right (thanks Fabian).

### Hexen
* Mouse buttons are now bindable to turn left/right (thanks Fabian).

### Strife
* Mouse buttons are now bindable to turn left/right (thanks Fabian).

```

</details>

Also, are these going to be mentioned?
1. https://github.com/chocolate-doom/chocolate-doom/pull/1660
2. https://github.com/chocolate-doom/chocolate-doom/pull/1652
3.  I'm not sure how to word these related items:

Scroll up to rfomin's message

> The current chocpkg system does not support FluidSynth. We can build a release for Windows using MSYS2 in GitHub CI.

Together with https://github.com/chocolate-doom/chocolate-doom/pull/1628:

> Added a build for macOS.
> Integrate MSYS2 build.
> Upload artifacts.

**Comment by rfomin** [comment]
> There was something about Windows builds not having up to date libraries. Has this been fixed? If not, what needs to be fixed?

>We'll need to update SDL2_Mixer to 2.28 in chocpkg.

How about this PR #1643? I think we should use GitHub CI for the release build instead of chocpkg.

Pluses:

* FluidSynth support for Windows and macOS
* SDL_Mixer 2.28 already included in [MSYS2](https://packages.msys2.org/base/mingw-w64-SDL2_mixer)
* 64-bit build on Windows.

Minuses:

* No Windows XP support

**Comment by fabiangreffrath** [comment]
Oh yes, forgot about it. Let's do this, please!

**Comment by mikeday0** [comment]
Does the `rpm.spec.in` file need attention? The stated dependencies are incorrect. Does anyone still use these? If not, maybe it should be removed.

**Comment by turol** [comment]
@mfrancis95 May I force push your branch?

**Comment by NY00123** [comment]
Just heard of a potential release becoming closer.

Hopefully, before that's done, we can have an answer on which variation of Hexen v1.1's behaviors to support.

As a reminder:
- At some point, I modified the released Hexen sources to match the most common executable known as "v1.1". It is also unofficially known as "1.1r1" nowadays. I submitted a PR for Chocolate Hexen and it was accepted.
- Only afterwards, I realized that, in fact, there's also an exe matching the released sources; It's just not commonly known or used. It got a similar (unofficial) title of "1.1r2".
- Therefore, I opened a related issue, and also submitted a PR aiming to re-introduce support for 1.1r2's behaviors in parallel to 1.1r1's.

The main question is, what is to be done for 3.1.0? Revert my change and behave like Chocolate Hexen 3.0.0-3.0.1 (i.e., Hexen 1.1r2), preserve it (1.1r1) or support both variations?

References:
https://doomwiki.org/w/index.php?title=HEXEN.EXE
https://github.com/chocolate-doom/chocolate-doom/pull/1394
https://github.com/chocolate-doom/chocolate-doom/issues/1378
https://github.com/chocolate-doom/chocolate-doom/pull/1266

**Comment by fabiangreffrath** [comment]
I'd say we should support the new `gameversion`, but not make it the default. Though, this is apparently exactly what happened when the alternative Final Doom version was introduced:

https://github.com/chocolate-doom/chocolate-doom/commit/ed4adeba77bf7dd803e7b944ee5f871965a0ca29

**Comment by NY00123** [comment]
> I'd say we should support the new `gameversion`, but not make it the default. Though, this is apparently exactly what happened when the alternative Final Doom version was introduced:
> 
> [ed4adeb](https://github.com/chocolate-doom/chocolate-doom/commit/ed4adeba77bf7dd803e7b944ee5f871965a0ca29)

Thanks for showing support for having both variations of Hexen 1.1 as options!

It seems that in Final Doom's case, the default was changed to the earlier version more than 6 months later, according to the timestamps of the converted SVN commits: https://github.com/chocolate-doom/chocolate-doom/commit/71d316afb2ae7191a4ef6fac2d757238ae3616e1

This still differs from the situation with Hexen, as Chocolate Hexen 3.0.1 and earlier should behave like "1.1r2", rather than "1.1r1".

The impacts should still be significantly smaller than in Doom. Apart from changing `HEXEN_VERSIONTEXT`, there are the `A_SoAExplode` fixes, impacting demo playback only if monsters are disabled. As in Doom v1.2, I don't think that the demo lumps themselves have a field for the "nomonsters" toggle.

I don't currently have much more to add on 3.1.0 in general. My hunch is that there's more than enough to cover, including but not limited to other improvements to accuracy during gameplay. After 3.1.0 is out, this might also become the standard version of Chocolate Doom for multiplayer, given that a revision differing from 3.0.0 and 3.0.1 has been suggested for a while, if I'm not wrong.

**Comment by fabiangreffrath** [comment]
We came a good way forward. Any other blockers left?

**Comment by turol** [comment]
Are all of today's PRs mentioned? In particular the Hexen/Strife desync?

Do we still want to wait for #1665?

**Comment by JNechaevsky** [comment]
Give me couple of hours please, I was going to do it today, but we have yet another moving of furniture. 🤯

**Comment by turol** [comment]
There were no objections so I rebased and partially squahed this. Are we still missing something?

**Comment by mikeday0** [comment]
Still need mention of the last handful of PRs:

* Fix possible Hexen and Strife desync (Michael Day).
* Heretic: Fix null backsector crash (Julia Nechaevskaya).
* Hexen: Add support for two different v1.1 variants through `-gameversion` argument (@NY00123).

I would like to do some Strife gamepad improvements but that can wait until after the release.

**Comment by rfomin** [comment]
We need to update the [RELEASE_NOTES.md](https://github.com/chocolate-doom/chocolate-doom/blob/master/RELEASE_NOTES.md) file for the release on GitHub. I suggest just copying the `## 3.1.0 (2024-??-??)` section from NEWS.md into it.

**Comment by turol** [comment]
@rfomin Updating RELEASE_NOTES.md should be done once just before release so it doesn't accidentally get left incomplete.

@NY00123 I couldn't find a PR for bash completion fix. Do you have it?

**Comment by NY00123** [comment]
> @NY00123 I couldn't find a PR for bash completion fix. Do you have it?

I can't find a clear relevant PR referencing bash completion, either. Maybe @fragglet will recall, being the one to list the fix in PR #1435.

Referencing possible commits:

b9d4c04c840321f5ec70787d8afb1256766aaa01
96e0123cd27a88a69c4582eb1e7c8dda66f3114d (probably not that one)
da88012088dad9f7a5d53c26aa40aa393cb5b703

**Comment by fabiangreffrath** [comment]
> Referencing possible commits:

Pretty sure the last one.

**Comment by turol** [comment]
If we're happy with the changelog then we should...
1. Pick a release date
2. I'll add it to the changelog and rebase one last time.
3. I'll copy the 3.1.0 stuff to RELEASE_NOTES.md
4. Merge this PR
5. The actual release. What needs to be done to make this happen? Just push a tag?

**Comment by rfomin** [comment]
> 5\. The actual release. What needs to be done to make this happen? Just push a tag?

Yes, just push a tag. Although it will only be a release on GitHub, not chocolate-doom.org.

**Comment by mikeday0** [comment]
> If we're happy with the changelog then we should...
> 
>     1. Pick a release date
> 

I am just a humble contributor but I vote next Monday - April Fools' Day. :)

**Comment by fabiangreffrath** [comment]
Another four months have passed, looks like we have lost track of this again. Could we please go forward and just decide on e.g. August 1st as a proper release date, please?

**Comment by turol** [comment]
Works for me.

**Review by fabiangreffrath — APPROVED** [review]
Let's go!

**Comment by turol** [comment]
I've drafted the new release. Project maintainers should be able to see it under releases. Please check it over. I will hit "Publish Release" sometime tomorrow afternoon.

**Pull request merged**
